app/src/main/java/nerd/tuxmobil/fahrplan/congress/preferences/AlarmTonePreference.kt

```diff
@@ -0,0 +1,78 @@
+package nerd.tuxmobil.fahrplan.congress.preferences
+
+import android.app.Activity
+import android.content.Context
+import android.content.Intent
+import android.media.RingtoneManager
+import android.net.Uri
+import android.provider.Settings
+import android.util.AttributeSet
+import androidx.preference.Preference
+import nerd.tuxmobil.fahrplan.congress.extensions.withExtras
+import nerd.tuxmobil.fahrplan.congress.repositories.AppRepository
+
+/**
+ * Represents an alarm tone preference.
+ */
+class AlarmTonePreference : Preference {
+
+    companion object {
+
+        /**
+         * See [Settings.System.DEFAULT_ALARM_ALERT_URI]
+         */
+        val DEFAULT_VALUE_URI: Uri = Settings.System.DEFAULT_ALARM_ALERT_URI
+
+        /**
+         * Content `Uri` string representation of [DEFAULT_VALUE_URI].
+         */
+        val DEFAULT_VALUE_STRING: String = DEFAULT_VALUE_URI.toString()
+
+        fun onAlarmTonePicked(appRepository: AppRepository, intent: Intent) {
+            val alarmToneUri = intent.getParcelableExtra<Uri>(RingtoneManager.EXTRA_RINGTONE_PICKED_URI)
+            appRepository.updateAlarmToneUri(alarmToneUri)
+        }
+
+        fun showAlarmTonePicker(appRepository: AppRepository, activity: Activity, requestCode: Int) {
+            val alarmToneUri = appRepository.readAlarmToneUri()
+            val intent = Intent(RingtoneManager.ACTION_RINGTONE_PICKER).withExtras(
+                    RingtoneManager.EXTRA_RINGTONE_TYPE to RingtoneManager.TYPE_ALARM,
+                    RingtoneManager.EXTRA_RINGTONE_SHOW_DEFAULT to true,
+                    RingtoneManager.EXTRA_RINGTONE_SHOW_SILENT to true,
+                    RingtoneManager.EXTRA_RINGTONE_DEFAULT_URI to DEFAULT_VALUE_URI,
+                    RingtoneManager.EXTRA_RINGTONE_EXISTING_URI to alarmToneUri
+            )
+            activity.startActivityForResult(intent, requestCode)
+        }
```

Comment:
The dependency on `AppRepository` kills the reusability of this class. You can access the persisted value using `getPersistedString()` and update it using `persistString()`.

Reply:
Thanks for the hint. I got as fas as this:

``` kotlin
override fun onClick() {
    super.onClick()
    val alarmToneUri = getPersistedString(DEFAULT_VALUE_STRING).toUri()
    val intent = Intent(RingtoneManager.ACTION_RINGTONE_PICKER).withExtras(
            RingtoneManager.EXTRA_RINGTONE_TYPE to RingtoneManager.TYPE_ALARM,
            RingtoneManager.EXTRA_RINGTONE_SHOW_DEFAULT to true,
            RingtoneManager.EXTRA_RINGTONE_SHOW_SILENT to true,
            RingtoneManager.EXTRA_RINGTONE_DEFAULT_URI to DEFAULT_VALUE_URI,
            RingtoneManager.EXTRA_RINGTONE_EXISTING_URI to alarmToneUri
    )
    context.startActivityForResult(intent, REQUEST_CODE_ALARM_TONE) // not possible here
}

override fun onActivityResult(requestCode: Int, resultCode: Int, intent: Intent?): Boolean {
    if (REQUEST_CODE_ALARM_TONE == requestCode && Activity.RESULT_OK == resultCode && intent != null) {
        val alarmToneUri = intent.getParcelableExtra<Uri>(RingtoneManager.EXTRA_RINGTONE_PICKED_URI)
        Logging.get().d(javaClass.simpleName, "Alarm tone picked: '$alarmToneUri'")
        return persistString(alarmToneUri.toString())
    }
    return false
}
```

I am missing accessing to an `Activity` to invoke `startActivityForResult`, though.

Reply:
You could go the hacky route and do

```kotlin
val activity = (preferenceManager.onDisplayPreferenceDialogListener as Fragment).activity
```

But hardcoding the request code severely limits reusability. So you might as well just add an `init(activity: Activity, requestCode: Int)` function.

Reply:
You could also do

```kotlin
override fun onClick() {
    preferenceManager.showDialog(this)
}
```

And in `SettingsFragment`:

```kotlin
override fun onDisplayPreferenceDialog(preference: Preference?) {
    if (preference is AlarmTonePreference) {
        preference.showAlarmTonePicker(requireActivity(), REQUEST_CODE_ALARM_TONE)
    } else {
        super.onDisplayPreferenceDialog(preference)
    }
}
```

Reply:
I changed it in 8a1cdab07ba57e017e94c5ead9cb6d2309773bc2 but now `onActivityResult` does not get called.

Reply:
My bad. You need to call `startActivityForResult()` on the fragment, not the activity.

Reply:
:heavy_check_mark: Fixed in 8f6cb562db48c32fa9ec957ed65fdf7f2c5f918a. Thx.